charms/istio-pilot/lib/charms/istio_pilot/v0/ingress.py

```diff
@@ -0,0 +1,268 @@
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""# Ingress Interface Library.
+"""
+
+import logging
+from functools import cached_property
+
+from ops.charm import CharmBase
+from ops.framework import EventBase, EventSource, Object, ObjectEvents
+from ops.model import BlockedStatus, WaitingStatus
+from serialized_data_interface import get_interface, NoCompatibleVersions, NoVersionsListed
+
+logger = logging.getLogger(__name__)
+
+# The unique Charmhub library identifier, never change it
+LIBID = "b521889515b34432b952f75c21e96dfc"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+DEFAULT_RELATION_NAME = "ingress"
+
+
+class IngressProviderAvailableEvent(EventBase):
+    """Event triggered when the ingress provider is ready for requests."""
+
+
+class IngressReadyEvent(EventBase):
+    """Event triggered when the ingress provider has returned the requested URL(s)."""
+
+
+class IngressFailedEvent(EventBase):
+    """Event triggered when something went wrong with the ingress relation."""
+
+
+class IngressRemovedEvent(EventBase):
+    """Event triggered when the ingress relation is removed."""
+
+
+class IngressRequirerEvents(ObjectEvents):
+    available = EventSource(IngressProviderAvailableEvent)
+    ready = EventSource(IngressReadyEvent)
+    failed = EventSource(IngressFailedEvent)
+    removed = EventSource(IngressRemovedEvent)
+
+
+class IngressRequirer(Object):
+    on = IngressRequirerEvents()
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str = DEFAULT_RELATION_NAME,
+        *,
+        port: int = None,
```

Comment:
I feel that, if unspecified, we should expose all the ports in the service

Reply:
There's no way to know that except by introspecting the K8s service directly, which goes against how relations are intended to work in Juju.  Besides, Istio only accepts a single port in the [`Destination`](https://istio.io/latest/docs/reference/config/networking/virtual-service/#Destination) specification, anyway (and requires it whenever a service does expose multiple ports, to disambiguate).

Reply:
To clarify on that last bit, Istio is mapping a single HTTP gateway + port to multiple backend services via paths. There's no way to do that for multiple ports unless you request a separate mapping (i.e. path) for each desired port. If multiple ports need to be exposed, this would be a use-case for a charm sending multiple ingress requests on a single relation.

Reply:
> There's no way to know that except by introspecting the K8s service directly, which goes against how relations are intended to work in Juju

I am failing to understand the "goes against how relations are intended to work in Juju". I do not see anything un-Juju in there tbh.

> Istio only accepts a single port in the Destination specification

OK, so if unspecified, let's raise if the service has more than one port :-)

Reply:
> I am failing to understand the "goes against how relations are intended to work in Juju". I do not see anything un-Juju in there tbh.

I mean that in theory the interface could be used outside of K8s for some other type of ingress provider / consumer, in which case there would be no way to "introspect the service" to determine the default port. Although, TBH, the `service` field already breaks that a bit; it should really be `destination` and go back to being required, since whether it's a service name (which is really a host name in K8s) or the ingress address would depend on what type of charm is using the relation.

Reply:
> I mean that in theory the interface could be used outside of K8s for some other type of ingress provider / consumer

In that case I would expect a library with a similar or same API, owned by a different charm.